lib/clamp-device.in

```diff
@@ -177,11 +177,17 @@ fi
 
 # Invoke HCC-specific opt passes
 if [ "$HCC_TILECHECK" == "ON" ]; then
-    $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
-        -select-accelerator-code -dce -globaldce -always-inline -tile_uniform < $2.linked.bc -o $2.selected.bc
+  $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+    -load $LIB/LLVMPromotePointerKernArgsToGlobal@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+    -select-accelerator-code -promote-pointer-kernargs-to-global \
+    -infer-address-spaces -dce -globaldce -always-inline -tile_uniform \
+    < $2.linked.bc -o $2.selected.bc
 else
-    $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
-        -select-accelerator-code -dce -globaldce -always-inline < $2.linked.bc -o $2.selected.bc
+  $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+    -load $LIB/LLVMPromotePointerKernArgsToGlobal@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+    -select-accelerator-code -promote-pointer-kernargs-to-global \
+    -infer-address-spaces -dce -globaldce -always-inline \
+    < $2.linked.bc -o $2.selected.bc
```

Comment:
I just want to confirm that we do inlining before promote-pointer-kernargs-to-global and infer-address-spaces because otherwise, we won't be getting the opportunities in the inlined callees.

Reply:
at this stage we do the following passes:
```
-targetlibinfo
-tti
-targetpassconfig
-assumption-cache-tracker
-profile-summary-info
-select-accelerator-code
-promote-pointer-kernargs-to-global
-dce
-globaldce
-basiccg
-always-inline
-verify
-write-bitcode
```

only `always-inline` is used not the full inliner. i think the optimization proposed in this PR is only for kernels, not device functions used by kernels?

Reply:
I think we want to run the PromotePointerKernArgs and inferAddressSpace passes after full inlining such that the inlined callee's code also gets exposed to the address space promotion optimization.